code-walktrhough.md

```diff
@@ -0,0 +1,206 @@
+
+While the RISC Zero zkVM enables developers to write off-chain logic as guest code, zkVM proofs (specifically a Groth16 SNARKs) get verified and incorporated into smart contract logic on chain. The end result is that the guest logic which checks if a number is even, is performed off chain but enforced on chain. This model extends to any program, however complex, which can be written as a guest program of the RISC Zero zkVM via Rust.
+The EvenNumber smart contract verifies a RISC Zero zkVM proof that a number is even.
+In this simple example, if a proof of an even number is verified, we will set that even number in the smart contract's state.
+
+### Key Components:
+- **RISC Zero Verifier**: An interface to the RISC Zero verifier contract, which checks the validity of a zero-knowledge proof. The actual verifier contract would be deployed independently, and its address is passed to the `EvenNumber` contract upon initialization.
+- **Image ID**: Identifier for a specific zkVM guest binary (ELF) that the contract accepts proofs from. The image ID is similar to the address of a smart contract. It uniquely represents the logic of that guest program, ensuring that only proofs generated from a pre-defined guest program (in this case, checking if a number is even) are considered valid.
+- **Initialization**: Sets up the contract with a RISC Zero verifier address.
+- **Set Function**: Updates the stored number after successful proof verification.
+- **Get Function**: Returns the currently stored even number.
+
+### Contract walkthrough:
+
+```solidity
+pragma solidity ^0.8.20;
+
+import {IRiscZeroVerifier} from "risc0/IRiscZeroVerifier.sol";
+import {ImageID} from "./ImageID.sol";
+
+contract EvenNumber {
+    IRiscZeroVerifier public immutable verifier;
+    bytes32 public constant imageId = ImageID.IS_EVEN_ID;
+    uint256 public number;
+
+    constructor(IRiscZeroVerifier _verifier) {
+        verifier = _verifier;
+        number = 0;
+    }
+
+    function set(uint256 x, bytes32 postStateDigest, bytes calldata seal) public {
+        bytes memory journal = abi.encode(x);
+        require(verifier.verify(seal, imageId, postStateDigest, sha256(journal)));
+        number = x;
+    }
+
+    function get() public view returns (uint256) {
+        return number;
+    }
+}
+```
+#### Constructor
+
+The constructor requires the address of the RISC Zero verifier contract as an argument. This address is then set to an immutable variable, ensuring that the verification process always refers to the specified verifier. Finally, initializes the number state variable to 0, representing the initial state of the stored number.
+
+#### Setting the Even Number
+
+The **set** function updates the stored number, but only if the provided SNARK proof validates that the number is indeed even.
+
+##### Inputs:
+- **x**: The new even number to store.
+- **postStateDigest**: A hash representing the expected state of the zkVM after executing the computation.
```

Comment:
Not for this PR, but I am really looking forward to when we can drop this from the developers' awareness 😅